File: backend/app/services/websocket_pool_manager.py
"""WebSocket connection pool manager for optimizing WebSocket connections across sessions.

This service manages WebSocket connections to reduce overhead when switching between
different sessions for the same agent.
"""

import asyncio
import uuid
from typing import Dict, List, Tuple, Optional, Set
from collections import defaultdict
from datetime import datetime, timedelta
from loguru import logger

from fastapi import WebSocket, WebSocketDisconnect


class WebSocketPoolManager:
    """Manages a pool of WebSocket connections with support for session reuse and connection health monitoring."""

    def __init__(self, connection_timeout: int = 300, heartbeat_interval: int = 30):
        # agent_id_str -> list of ConnectionInfo tuples (maintaining compatibility with old API)
        self.active_connections: Dict[str, List[Tuple[WebSocket, Optional[str], Set[str]]]] = defaultdict(list)
        self._connection_timeout = connection_timeout  # seconds
        self._heartbeat_interval = heartbeat_interval  # seconds
        self._health_check_task = None
        self._connection_metadata: Dict[WebSocket, Dict] = {}  # Additional metadata for connections

    async def start_health_monitoring(self):
        """Start background task for connection health monitoring."""
        if self._health_check_task is None:
            self._health_check_task = asyncio.create_task(self._connection_health_loop())

    async def stop_health_monitoring(self):
        """Stop the background health monitoring task."""
        if self._health_check_task:
            self._health_check_task.cancel()
            try:
                await self._health_check_task
            except asyncio.CancelledError:
                pass
            self._health_check_task = None

    async def _connection_health_loop(self):
        """Background task to monitor connection health and cleanup stale connections."""
        while True:
            try:
                await asyncio.sleep(self._heartbeat_interval)
                await self.cleanup_stale_connections()
            except asyncio.CancelledError:
                logger.info("WebSocket pool health monitoring stopped")
                break
            except Exception as e:
                logger.error(f"Error in WebSocket pool health monitoring: {e}")

    def get_connection_count(self, agent_id: str) -> int:
        """Get the total number of connections for an agent."""
        return len(self.active_connections.get(agent_id, []))

    def get_session_count(self, agent_id: str) -> int:
        """Get the total number of sessions across all connections for an agent."""
        count = 0
        for ws, current_session_id, session_set in self.active_connections.get(agent_id, []):
            count += len(session_set)
        return count

    async def connect(self, agent_id: str, websocket: WebSocket, session_id: str = None):
        """Connect a WebSocket to the pool, potentially reusing an existing connection."""
        await websocket.accept()

        # For backward compatibility, check if there's an existing connection that can be reused
        # but keep the same signature as the original ConnectionManager

        # If no existing connection for this agent, create the initial structure
        if agent_id not in self.active_connections:
            self.active_connections[agent_id] = []

        # Add the new connection with its session ID and set of associated sessions
        session_set = {session_id} if session_id else set()

        # Check if there's an existing connection for this agent that we can reuse
        reuse_found = False
        for i, (ws, current_sid, sess_set) in enumerate(self.active_connections[agent_id]):
            if ws == websocket:
                # This shouldn't happen in normal flow, but just in case
                self.active_connections[agent_id][i] = (websocket, session_id, sess_set)
                reuse_found = True
                break
            elif ws.application_state == 1 and session_id and session_id not in sess_set:
                # Reuse an existing active connection that doesn't already have this session
                # Add the session to the existing set and update
                new_sess_set = sess_set.copy()
                new_sess_set.add(session_id)
                self.active_connections[agent_id][i] = (ws, session_id, new_sess_set)
                reuse_found = True
                break

        # If no suitable connection was found to reuse, add as a new connection
        if not reuse_found:
            self.active_connections[agent_id].append((websocket, session_id, session_set))

        # Store connection metadata
        self._connection_metadata[websocket] = {
            'created_at': datetime.now(),
            'last_activity': datetime.now(),
            'agent_id': agent_id
        }

    def disconnect(self, agent_id: str, websocket: WebSocket):
        """Remove a WebSocket from the pool."""
        if agent_id in self.active_connections:
            # Remove the specific connection
            self.active_connections[agent_id] = [
                (ws, sid, sess_set) for ws, sid, sess_set in self.active_connections[agent_id]
                if ws != websocket
            ]

            # Clean up metadata
            if websocket in self._connection_metadata:
                del self._connection_metadata[websocket]

            # Clean up agent entry if no connections remain
            if not self.active_connections[agent_id]:
                del self.active_connections[agent_id]

    def switch_session(self, agent_id: str, websocket: WebSocket, new_session_id: str) -> bool:
        """Switch the active session for a WebSocket connection."""
        if agent_id in self.active_connections:
            for i, (ws, current_sid, sess_set) in enumerate(self.active_connections[agent_id]):
                if ws == websocket:
                    # Add the new session to the set of sessions for this connection
                    new_sess_set = sess_set.copy()
                    new_sess_set.add(new_session_id)
                    # Update the current session for this connection
                    self.active_connections[agent_id][i] = (ws, new_session_id, new_sess_set)

                    # Update metadata
                    if websocket in self._connection_metadata:
                        self._connection_metadata[websocket]['last_activity'] = datetime.now()

                    return True
        return False

    async def send_message(self, agent_id: str, message: dict, session_id: str = None):
        """Send message to connections associated with the agent and optionally specific session."""
        if agent_id in self.active_connections:
            for ws, current_sid, sess_set in self.active_connections[agent_id]:
                # If session_id is specified, only send to connections that include this session
                if session_id is None or session_id in sess_set:
                    try:
                        await ws.send_json(message)

                        # Update activity timestamp
                        if ws in self._connection_metadata:
                            self._connection_metadata[ws]['last_activity'] = datetime.now()
                    except WebSocketDisconnect:
                        # Remove disconnected connection
                        self.disconnect(agent_id, ws)
                        continue
                    except Exception as e:
                        logger.error(f"Error sending message to WebSocket: {e}")

    def get_active_session_ids(self, agent_id: str) -> List[str]:
        """Return distinct session IDs for all active WS connections of an agent."""
        if agent_id not in self.active_connections:
            return []
        all_sessions = set()
        for _ws, _current_sid, sess_set in self.active_connections[agent_id]:
            all_sessions.update(sess_set)
        return list(all_sessions)

    def get_connection_for_session(self, agent_id: str, session_id: str) -> Optional[WebSocket]:
        """Find the WebSocket connection that handles the given session."""
        if agent_id in self.active_connections:
            for ws, current_sid, sess_set in self.active_connections[agent_id]:
                if session_id in sess_set:
                    return ws
        return None

    async def cleanup_stale_connections(self):
        """Clean up connections that have been inactive beyond the timeout."""
        current_time = datetime.now()
        agents_to_remove = []

        for agent_id, connections in self.active_connections.items():
            active_connections = []
            for ws, current_sid, sess_set in connections:
                # Check if WebSocket is still connected and get metadata
                is_connected = ws.application_state == 1  # WebSocketState.CONNECTED
                last_activity = self._connection_metadata.get(ws, {}).get('last_activity', current_time)

                time_since_activity = (current_time - last_activity).seconds

                if is_connected and time_since_activity < self._connection_timeout:
                    active_connections.append((ws, current_sid, sess_set))
                else:
                    # Remove disconnected or stale connections
                    try:
                        if not is_connected:
                            logger.info(f"Removing disconnected WebSocket for agent {agent_id}")
                        else:
                            logger.info(f"Removing stale WebSocket for agent {agent_id} (inactive for {time_since_activity}s)")
                        await ws.close(code=1000)  # Normal closure
                    except:
                        pass  # Ignore errors during close

            if active_connections:
                self.active_connections[agent_id] = active_connections
            else:
                agents_to_remove.append(agent_id)

        # Remove agents with no active connections
        for agent_id in agents_to_remove:
            if agent_id in self.active_connections:
                del self.active_connections[agent_id]
            # Clean up any remaining metadata for connections from this agent
            connections_to_clean = []
            for ws, meta in self._connection_metadata.items():
                if meta.get('agent_id') == agent_id:
                    connections_to_clean.append(ws)
            for ws in connections_to_clean:
                del self._connection_metadata[ws]

File: frontend/src/services/WebSocketPoolManager.ts
/**
 * WebSocketPoolManager - Manages WebSocket connections for different agents and sessions
 * to optimize connection reuse and reduce overhead when switching between sessions.
 */

interface WebSocketConnection {
  ws: WebSocket;
  agentId: string;
  activeSessionId: string | null;
  associatedSessions: Set<string>;
  createdAt: Date;
  lastActivity: Date;
  isConnected: boolean;
}

class WebSocketPoolManager {
  private connections: Map<string, WebSocketConnection> = new Map();
  private connectionTimeout: number = 300000; // 5 minutes in milliseconds
  private heartbeatInterval: number = 30000; // 30 seconds in milliseconds
  private healthCheckInterval: NodeJS.Timeout | null = null;

  constructor(timeoutMs: number = 300000, heartbeatMs: number = 30000) {
    this.connectionTimeout = timeoutMs;
    this.heartbeatInterval = heartbeatMs;
  }

  /**
   * Start the health monitoring for connections
   */
  startHealthMonitoring(): void {
    if (this.healthCheckInterval) {
      clearInterval(this.healthCheckInterval);
    }

    this.healthCheckInterval = setInterval(() => {
      this.cleanupStaleConnections();
    }, this.heartbeatInterval);
  }

  /**
   * Stop the health monitoring for connections
   */
  stopHealthMonitoring(): void {
    if (this.healthCheckInterval) {
      clearInterval(this.healthCheckInterval);
      this.healthCheckInterval = null;
    }
  }

  /**
   * Get or create a WebSocket connection for the specified agent
   */
  getOrCreateConnection(agentId: string, token: string, sessionParam: string = ''): WebSocketConnection {
    // Look for an existing connection for this agent that is still active
    for (const [key, conn] of this.connections.entries()) {
      if (conn.agentId === agentId &&
          conn.isConnected &&
          (Date.now() - conn.lastActivity.getTime()) < this.connectionTimeout) {

        // Reuse existing connection
        conn.lastActivity = new Date();
        return conn;
      }
    }

    // Create new connection
    const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
    const wsUrl = `${protocol}//${window.location.host}/ws/chat/${agentId}?token=${token}${sessionParam}`;
    const ws = new WebSocket(wsUrl);

    const connectionId = `${agentId}-${Date.now()}`;
    const connection: WebSocketConnection = {
      ws,
      agentId,
      activeSessionId: sessionParam.includes('session_id=')
        ? new URLSearchParams(sessionParam.substring(1)).get('session_id')
        : null,
      associatedSessions: sessionParam.includes('session_id=')
        ? new Set([new URLSearchParams(sessionParam.substring(1)).get('session_id')!])
        : new Set(),
      createdAt: new Date(),
      lastActivity: new Date(),
      isConnected: false
    };

    // Set up connection event handlers
    ws.onopen = () => {
      connection.isConnected = true;
      connection.lastActivity = new Date();
    };

    ws.onclose = () => {
      connection.isConnected = false;
      this.removeConnection(connectionId);
    };

    ws.onerror = () => {
      connection.lastActivity = new Date();
    };

    ws.onmessage = (event) => {
      connection.lastActivity = new Date();
    };

    this.connections.set(connectionId, connection);
    return connection;
  }

  /**
   * Switch the active session for an existing WebSocket connection
   */
  switchSession(ws: WebSocket, newSessionId: string): boolean {
    for (const [_, conn] of this.connections.entries()) {
      if (conn.ws === ws) {
        // Add the new session to the set of associated sessions
        conn.associatedSessions.add(newSessionId);
        // Update the active session
        conn.activeSessionId = newSessionId;
        conn.lastActivity = new Date();

        // Send session switch message if connection is open
        if (conn.isConnected && ws.readyState === WebSocket.OPEN) {
          ws.send(JSON.stringify({
            type: 'switch_session',
            session_id: newSessionId
          }));
        }
        return true;
      }
    }
    return false;
  }

  /**
   * Send a message through the WebSocket connection for a specific agent/session
   */
  sendMessage(agentId: string, message: any, sessionId?: string): boolean {
    for (const [_, conn] of this.connections.entries()) {
      if (conn.agentId === agentId &&
          (sessionId ? conn.associatedSessions.has(sessionId) : true) &&
          conn.isConnected &&
          conn.ws.readyState === WebSocket.OPEN) {

        conn.ws.send(JSON.stringify(message));
        conn.lastActivity = new Date();
        return true;
      }
    }
    return false;
  }

  /**
   * Get the WebSocket for a specific agent and session
   */
  getConnectionForSession(agentId: string, sessionId: string): WebSocket | null {
    for (const [_, conn] of this.connections.entries()) {
      if (conn.agentId === agentId &&
          conn.associatedSessions.has(sessionId) &&
          conn.isConnected) {
        return conn.ws;
      }
    }
    return null;
  }

  /**
   * Get all active session IDs for a specific agent
   */
  getActiveSessionIds(agentId: string): string[] {
    const sessionIds: string[] = [];
    for (const [_, conn] of this.connections.entries()) {
      if (conn.agentId === agentId) {
        sessionIds.push(...Array.from(conn.associatedSessions));
      }
    }
    return [...new Set(sessionIds)]; // Remove duplicates
  }

  /**
   * Get the number of connections for an agent
   */
  getConnectionCount(agentId: string): number {
    let count = 0;
    for (const [_, conn] of this.connections.entries()) {
      if (conn.agentId === agentId && conn.isConnected) {
        count++;
      }
    }
    return count;
  }

  /**
   * Close and remove a specific connection
   */
  removeConnection(connectionId: string): void {
    const connection = this.connections.get(connectionId);
    if (connection) {
      if (connection.ws.readyState === WebSocket.OPEN ||
          connection.ws.readyState === WebSocket.CONNECTING) {
        connection.ws.close();
      }
      this.connections.delete(connectionId);
    }
  }

  /**
   * Close all connections for a specific agent
   */
  closeAgentConnections(agentId: string): void {
    for (const [connectionId, conn] of this.connections.entries()) {
      if (conn.agentId === agentId) {
        if (conn.ws.readyState === WebSocket.OPEN ||
            conn.ws.readyState === WebSocket.CONNECTING) {
          conn.ws.close();
        }
        this.connections.delete(connectionId);
      }
    }
  }

  /**
   * Close all connections in the pool
   */
  closeAllConnections(): void {
    if (this.healthCheckInterval) {
      clearInterval(this.healthCheckInterval);
      this.healthCheckInterval = null;
    }

    for (const [_, conn] of this.connections.entries()) {
      if (conn.ws.readyState === WebSocket.OPEN ||
          conn.ws.readyState === WebSocket.CONNECTING) {
        conn.ws.close();
      }
    }
    this.connections.clear();
  }

  /**
   * Clean up stale connections that have been inactive
   */
  private cleanupStaleConnections(): void {
    const now = Date.now();
    for (const [connectionId, conn] of this.connections.entries()) {
      const timeSinceActivity = now - conn.lastActivity.getTime();

      // Check if connection is stale (inactive for too long)
      if (timeSinceActivity > this.connectionTimeout) {
        console.log(`Cleaning up stale connection for agent ${conn.agentId}`);
        this.removeConnection(connectionId);
      }
      // Check if WebSocket is actually closed but we haven't been notified yet
      else if (conn.ws.readyState === WebSocket.CLOSED) {
        this.removeConnection(connectionId);
      }
    }
  }

  /**
   * Check if a connection exists for an agent
   */
  hasActiveConnection(agentId: string): boolean {
    for (const [_, conn] of this.connections.entries()) {
      if (conn.agentId === agentId && conn.isConnected) {
        return true;
      }
    }
    return false;
  }
}

// Create a singleton instance
const wsPoolManager = new WebSocketPoolManager();

export default wsPoolManager;